File: apps/dotnet/src/MachineLearning/Chapter08/WeightedTree.fs
module MachineLearning.Chapter08.WeightedTree

open MachineLearning.Chapter03.DecisionTree

/// ラベルごとの重みの合計。ラベルが最初に現れた順に並ぶ
let private sumWeightsByLabel (labels: 'L list) (weights: float list) : ('L * float) list =
    List.zip labels weights
    |> List.groupBy fst
    |> List.map (fun (label, pairs) -> label, List.sumBy snd pairs)

/// 重み付きのジニ不純度。ラベルの件数の代わりに重みの合計で割合を求める
let weightedGini (labels: 'L list) (weights: float list) : float =
    let total = List.sum weights

    1.0
    - (sumWeightsByLabel labels weights
       |> List.sumBy (fun (_, weight) -> (weight / total) ** 2.0))

/// クラスの重みの付け方
type ClassWeight =
    /// すべての行の重みを 1 にする
    | Unweighted
    /// 1 件の重みを「件数 / (クラスの数 × そのクラスの件数)」にして、クラスごとの重みの合計をそろえる
    | Balanced

/// 決定木の学習の設定
type TreeOptions =
    {
        /// 木の深さの上限。None なら制限しない
        MaxDepth: int option
        ClassWeight: ClassWeight
    }

/// 正解ラベルから、1 件ごとの重みを求める
let classWeights (classWeight: ClassWeight) (t: 'L list) : float list =
    match classWeight with
    | Unweighted -> t |> List.map (fun _ -> 1.0)
    | Balanced ->
        let counts = t |> List.countBy id |> Map.ofList

        t
        |> List.map (fun label -> float t.Length / float (counts.Count * counts[label]))

/// 1 つの特徴量について、隣り合う値の中点を境界の候補にした分け方。不純度は重みの合計で重み付けする
let private splitsOf (feature: string) (x: Map<string, float> list) (t: 'L list) (weights: float list) : Split list =
    let rows =
        List.zip3 (x |> List.map (fun row -> row[feature])) t weights
        |> List.sortBy (fun (value, _, _) -> value)

    let valueAt i =
        let value, _, _ = rows[i]
        value

    let total = List.sum weights

    [ 1 .. rows.Length - 1 ]
    |> List.filter (fun i -> valueAt (i - 1) <> valueAt i)
    |> List.map (fun i ->
        let left, right = List.splitAt i rows

        let impurityOf (part: (float * 'L * float) list) =
            let labels = part |> List.map (fun (_, label, _) -> label)
            let partWeights = part |> List.map (fun (_, _, weight) -> weight)
            List.sum partWeights * weightedGini labels partWeights

        {
            Feature = feature
            Threshold = (valueAt (i - 1) + valueAt i) / 2.0
            Impurity = (impurityOf left + impurityOf right) / total
        })

/// 重み付きの不純度が最も小さくなる分け方。ラベルが 1 種類か、分けられる値が無ければ None
let bestWeightedSplit (x: Map<string, float> list) (t: 'L list) (weights: float list) : Split option =
    match x with
    | [] -> None
    | _ when weightedGini t weights = 0.0 -> None
    | first :: _ ->
        first
        |> Map.keys
        |> Seq.toList
        |> List.collect (fun feature -> splitsOf feature x t weights)
        |> function
            | [] -> None
            | splits -> Some(List.minBy (fun split -> split.Impurity) splits)

/// 重みの合計が最も大きいラベル。同数なら先に現れたラベルを選ぶ
let weightedMajority (labels: 'L list) (weights: float list) : 'L =
    sumWeightsByLabel labels weights |> List.maxBy snd |> fst

/// 第 3 章の fit と同じ手順で、1 件ごとの重みを通して木を作る。木の型は第 3 章の Tree<'L> をそのまま使う
// 左右の部分木を作ってから Node にまとめるので末尾再帰ではない。再帰の深さは木の深さまで
// fsharplint:disable-next-line EnsureTailCallDiagnosticsInRecursiveFunctions
let rec fitWeighted (maxDepth: int option) (x: Map<string, float> list) (t: 'L list) (weights: float list) : Tree<'L> =
    let split =
        if maxDepth = Some 0 then
            None
        else
            bestWeightedSplit x t weights

    match split with
    | None -> Leaf(weightedMajority t weights)
    | Some split ->
        let goesLeft (row: Map<string, float>, _, _) = row[split.Feature] <= split.Threshold
        let left, right = List.zip3 x t weights |> List.partition goesLeft
        let childDepth = maxDepth |> Option.map (fun depth -> depth - 1)

        let fitPart part =
            let partX, partT, partWeights = List.unzip3 part
            fitWeighted childDepth partX partT partWeights

        Node(split, fitPart left, fitPart right)

File: apps/ruby/lib/getting_started_ml/chapter08/weighted_tree.rb
# frozen_string_literal: true

module GettingStartedMl
  # 第 8 章のクラスの重みを付けた決定木。第 3 章の決定木に 1 件ごとの重みを足したもの。
  # 木の型（Leaf・Node・Split）と予測は第 3 章のものをそのまま使う。ラベルが整数でも型を変えずに済む。
  module Chapter08
    # クラスの重みの付け方。
    CLASS_WEIGHTS = %i[none balanced].freeze

    module_function

    # ラベルごとの重みの合計。Hash は先に現れたラベルの順を保つ。
    def weight_sums(labels, weights)
      labels.zip(weights).each_with_object(Hash.new(0.0)) { |(label, weight), sums| sums[label] += weight }
    end

    # 重み付きのジニ不純度。ラベルの件数の代わりに重みの合計で割合を求める。
    def weighted_gini(labels, weights)
      total = weights.sum
      return 0.0 if total.zero?

      1.0 - weight_sums(labels, weights).values.sum { |weight| (weight / total)**2 }
    end

    # クラスの件数に反比例する重み（件数 ÷（クラスの数 × そのクラスの件数））を 1 件ごとに求める。
    def balanced_weights(t)
      counts = t.tally
      t.map { |label| t.size.fdiv(counts.size * counts[label]) }
    end

    # クラスの重みの付け方から、1 件ごとの重みを求める。
    def weights_of(t, class_weight)
      case class_weight
      when :none then Array.new(t.size, 1.0)
      when :balanced then balanced_weights(t)
      else raise ArgumentError, "クラスの重みの付け方が違います: #{class_weight}"
      end
    end

    # 重みの合計が最も大きいラベル。同じなら先に現れたラベルを選ぶ。
    def weighted_majority(labels, weights)
      weight_sums(labels, weights).max_by { |_label, weight| weight }.first
    end

    # 左右の重み付き不純度の、重みによる平均が最も小さくなる分割を返す。分けられなければ nil。
    def best_split(x, t, weights)
      return nil if x.empty? || weighted_gini(t, weights).zero?

      x.first.columns.flat_map { |feature| candidates(x, t.zip(weights), feature) }.min_by(&:impurity)
    end

    # 1 つの列で、隣り合う値の中点を境界にした分割の候補をすべて返す。
    # 第 3 章の sort_by_feature に（ラベル, 重み）の組を渡して、重みも一緒に並べ替える。
    def candidates(x, labeled, feature)
      sorted = Chapter03.sort_by_feature(x, labeled, feature)

      (1...sorted.size).filter_map { |index| split_at(sorted, index, feature) }
    end

    # 並べた組の index 番目の手前で分けた分割を返す。前後の値が同じなら nil。
    def split_at(sorted, index, feature)
      before, after = sorted.values_at(index - 1, index).map(&:first)
      return nil if before == after

      labels, weights = sorted.map(&:last).transpose
      Chapter03::Split.new(feature:, threshold: (before + after) / 2.0,
                           impurity: split_impurity(labels, weights, index))
    end

    # index の手前と後ろに分けたときの、重みによる不純度の平均。
    def split_impurity(labels, weights, index)
      left = weights.first(index)
      right = weights.drop(index)

      ((left.sum * weighted_gini(labels.first(index), left)) +
        (right.sum * weighted_gini(labels.drop(index), right))) / weights.sum
    end

    # 深さの上限まで分割を繰り返して木を作る。max_depth が nil なら上限なし。
    def build(x, t, weights, max_depth)
      split = max_depth&.zero? ? nil : best_split(x, t, weights)
      return Chapter03::Leaf.new(label: weighted_majority(t, weights)) if split.nil?

      left, right = x.zip(t, weights).partition { |features, _, _| Chapter03.goes_left?(split, features) }
      Chapter03::Node.new(split:, left: build(*left.transpose, max_depth&.pred),
                          right: build(*right.transpose, max_depth&.pred))
    end

    # クラスの重みを付けられる決定木の分類器。fit で学習してから predict で予測する。
    class DecisionTreeClassifier
      attr_reader :tree

      def initialize(max_depth: nil, class_weight: :none)
        @max_depth = max_depth
        @class_weight = class_weight
        @tree = nil
      end

      # 訓練データから木を作る。メソッドをつなげられるように自分を返す。
      def fit(x, t)
        Chapter07.check_size(x.size, t.size)
        @tree = Chapter08.build(x, t, Chapter08.weights_of(t, @class_weight), @max_depth)
        self
      end

      # 特徴量ごとのラベルを予測する。
      def predict(x)
        raise "学習してから予測してください" if tree.nil?

        x.map { |features| Chapter03.predict_one(tree, features) }
      end

      def ==(other)
        other.is_a?(DecisionTreeClassifier) && tree == other.tree
      end
    end
  end
end
